File: day-17/conway_cubes.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

const input_file string = "./day-17/input.txt"

const ACTIVE = '#'

type Point struct {
	x, y, z, w int
	dimension  int
}

func (p Point) add(d []int) Point {
	// Pad out with zeros if not 4-dimensions
	for i := len(d); i < 4; i++ {
		d = append(d, 0)
	}
	return Point{p.x + d[0], p.y + d[1], p.z + d[2], p.w + d[3], p.dimension}
}

func (p Point) GetNeighbours() map[Point]bool {
	neighbours := make(map[Point]bool)
	for _, delta := range GetDeltas(p.dimension) {
		neighbours[p.add(delta)] = true
	}
	return neighbours
}

func GetRangeOfPoints(dimension int) [][]int {
	deltas := [][]int{}
	if dimension == 1 {
		return [][]int{{0}, {-1}, {1}}
	}
	for _, v := range []int{0, -1, 1} {
		for _, p := range GetRangeOfPoints(dimension - 1) {
			p = append(p, v)
			deltas = append(deltas, p)
		}
	}
	return deltas
}

func GetDeltas(dimension int) [][]int {
	deltas := GetRangeOfPoints(dimension)
	// Ignore all zeros. TODO - do this better
	return deltas[1:]
}

func ParseInputSet(input string, dimension int) map[Point]bool {
	active := make(map[Point]bool)
	for y, line := range strings.Split(strings.TrimSpace(input), "\n") {
		for x, r := range line {
			if r == ACTIVE {
				active[Point{x, y, 0, 0, dimension}] = true
			}
		}
	}
	return active
}

func CountNeighbours(point Point, active map[Point]bool) int {
	neighbourCount := 0
	for neighbour, _ := range point.GetNeighbours() {
		if _, found := active[neighbour]; found {
			neighbourCount++
		}
	}
	return neighbourCount
}

func Evolve(input string, iteration int, dimension int) int {
	active := ParseInputSet(input, dimension)
	for i := 0; i < iteration; i++ {
		pointsToCheck := make(map[Point]bool)
		for point, _ := range active {
			pointsToCheck[point] = true
			for neighbour, _ := range point.GetNeighbours() {
				pointsToCheck[neighbour] = true
			}
		}
		nextGen := make(map[Point]bool)
		for point, _ := range pointsToCheck {
			neighbourCount := CountNeighbours(point, active)
			if _, found := active[point]; found {
				if neighbourCount == 2 {
					nextGen[point] = true
				}
			}
			if neighbourCount == 3 {
				nextGen[point] = true
			}
		}
		active = nextGen
	}
	return len(active)
}

func Part1(input string, iterations int) int {
	return Evolve(input, iterations, 3)
}

func Part2(input string) int {
	return Evolve(input, 6, 4)
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data, 6))
	fmt.Printf("Part 2: %d\n", Part2(data))
}

File: day-17/conway_cubes.py
import itertools

input_file = 'day-17/input.txt'

class Point:
  def __init__(self, *args):
    self.coords = tuple(args)
    self.dimension = len(self.coords)
  
  def get_neighbours(self):
    neighbours = set()
    for delta in itertools.product([-1,0,1], repeat = self.dimension):
      neighbour = []
      for i in range(self.dimension):
        neighbour.append(self.coords[i] + delta[i])
      if tuple(neighbour) != self.coords:  
        neighbours.add(Point(*neighbour))
    return neighbours

  def __eq__(self, other):
    return isinstance(other, self.__class__) and self.coords == other.coords

  def __ne__(self, other):
    return not self.__eq__(other)

  def __hash__(self):
    return hash(self.coords)

  def __repr__(self):
    return f'{self.coords}'

def create_input_set(input, dimensions):
  active = set()
  lines = input.splitlines()
  for y in range(len(lines)):
    for x in range(len(lines[y])):
      if lines[y][x] == '#':
        active.add(Point(x,y,*([0] * (dimensions - 2))))
  return active

def count_neighbours(active, point):
  neighbour_count = 0
  for neighbour in point.get_neighbours():
    if neighbour in active:
      neighbour_count += 1
  return neighbour_count

def evolve(input, iterations, dimensions):
  active = create_input_set(input, dimensions)
  for i in range(iterations):
    points_to_check = set()
    for point in active:
      points_to_check = points_to_check.union({point})
      points_to_check = points_to_check.union(point.get_neighbours())
    next_gen = set()
    neighbour_cache = {}
    for point in points_to_check:
      neighbour_count = count_neighbours(active, point)
      if point in active and neighbour_count in [2,3]:
        next_gen.add(point)
      elif point not in active and neighbour_count == 3:
        next_gen.add(point)
    active = next_gen
  return len(active)

def part1(input, iterations = 6):
  return evolve(input, iterations, 3)

def part2(input, iterations = 6):
  return evolve(input, iterations, 4)

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))